receiver/mongodbreceiver/extractor.go

```diff
@@ -0,0 +1,437 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
+)
+
+type numberType int
+
+const (
+	integer numberType = iota
+	double
+)
+
+type mongoMetric struct {
+	name             string
+	path             []string
+	staticAttributes map[string]string
+	dataPointType    numberType
+	minMongoVersion  *version.Version
+	maxMongoVersion  *version.Version
+	conversionFactor int
+}
+
+// convertToAppropriateValue allows for definitions of metrics to be scaled to appropriate units,
+// example Megabytes to Bytes etc.
+func (mm *mongoMetric) convertToAppropriateValue(value interface{}) interface{} {
+	if mm.conversionFactor == 0 {
+		return value
+	}
+
+	switch mm.dataPointType {
+	case integer:
+		v, _ := value.(int64)
+		return v * int64(mm.conversionFactor)
+	case double:
+		v, _ := value.(float64)
+		return v * float64(mm.conversionFactor)
+	}
+
+	return value
+}
+
+const (
+	mebibytesToBytes = 1048576
+)
+
+type extractor struct {
+	version *version.Version
+	logger  *zap.Logger
+}
```

Comment:
I'm finding it very difficult to peel back all the layers of fancy generalization that this thing can do. Maybe this kind of thing makes sense eventually but the current scraper could be so much simpler to read & maintain if you do everything much more literally. "Dumb" code is ok. We don't want each scraper to be a unique work of art.


```
func scrape() {
	version := client.getVersion()

	dbs, err := client.getDBs()
	for db := range dbs {
		dbStats, err := client.DBStats(ctx, databaseName)
		if err != nil { ... }

		one, err := dig([]{"foo", "bar"}, dbStats)
		if err != nil { ... }
		i, err := parseInt(one)
		if err != nil { ... }
		mb.recordMetricOne(i)

		if version > 4.5 {
			two, err := dig([]{"foo", "bar"}, dbStats)
			if err != nil { ... }
			i, err := parseInt(v)
			if err != nil { ... }
			mb.recordMetricTwo(i)
		}
		
		...
	}
}
```

Fwiw, `extractAdminStats` is a lot more like how a scraper should look.